### PR TITLE
Feature match correction ui

### PR DESCRIPTION
**Pull request by opadron**
Removes the multi-tabbed interface in the feature-match workflow and replaces it with the in-browser correction UI (now, in a separate analysis page).

Updates the feature-match task script so that it returns all results; not just the results that matched, and adds the correction UI that uses the new results to present a UI that allows users to manually update/correct the computed match result by dragging and dropping.

### CODE REVIEW

File: server/task_specs/feature-match.py [hunk 1]
```diff
@@ -40,22 +40,65 @@ def compute_assignments(list1, list2, metric=None):
...
+    n2 = len(list2)
+    n_min = min(n1, n2)
+
+    # no more than *one* of these branches may be executed
+    left_over = None
+    if n1 != n_min:
+        left_over = np.empty((n1,), dtype=np.bool)
+        target = assign1
+
+    if n2 != n_min:
```
Comment (waxlamp) [inline server/task_specs/feature-match.py]: Can you use `elif` here? If the comment above is accurate, `elif` would enforce the "at most one branch" constraint.

File: server/task_specs/feature-match.py [hunk 2]
```diff
@@ -40,22 +40,65 @@ def compute_assignments(list1, list2, metric=None):
...
+        left_over = np.empty((n2,), dtype=np.bool)
+        target = assign2
+
+    if left_over is not None:
```
Comment (waxlamp) [inline server/task_specs/feature-match.py]: See `elif` comment above.

Comment (waxlamp) [reply server/task_specs/feature-match.py]: Or not. This should be `if` to check the state of `left_over` right?

File: server/ui_specs/feature-match-correction.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (waxlamp) [inline server/ui_specs/feature-match-correction.yml]: Blank line

File: server/ui_specs/feature-match-correction.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+---
+
```
Comment (waxlamp) [inline server/ui_specs/feature-match-correction.yml]: Blank line

File: web_client/analysis-modules/feature-matching-correction.jsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+import actions from '../actions';
+import analysisUtils from '../utils/analysis';
+import { Promise } from '../utils/promise';
+import { store } from '../globals';
+import loadModel from '../utils/load-model';
+
+import ItemModel from 'girder/models/ItemModel';
+import { formatSize } from 'girder/misc';
+
+const D = store.dispatch.bind(store);
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching-correction.jsx]: Don't use capitalized variables for things that are not classes.

File: web_client/analysis-modules/feature-matching-correction.jsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    let b = originalAssignments[i1].pair[1];
+
+    return [a, b];
+  })
+  .filter(([a, b]) => (!isNil(a) && !isNil(b)))
+  .map((pair) => ({ pair }));
+
+  const metaDataPromise = Promise.map(
+    priorData.inputFiles,
+    (input) => (
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching-correction.jsx]: Remove parens.

File: web_client/analysis-modules/feature-matching-correction.jsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  const { states } = data;
+  const outputs = {
+    output_path_1: `FILE:x:${states[priorData.page1Elements[0].id].name}`,
+    output_path_2: `FILE:x:${states[priorData.page1Elements[1].id].name}`
+  };
+
+  const title = 'feature-apply';
+  const maxPolls = 40;
+
+  const runPromise = analysisUtils.runTask(
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching-correction.jsx]: Unbreak line.

File: web_client/analysis-modules/feature-matching-correction.jsx [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  return Promise.all([removeResultElements, metaDataPromise, runPromise])
+    .then(([, meta, result]) => {
+      let { outputId1, outputId2 } = result;
+      return Promise.all([
+        loadModel(outputId1, ItemModel),
+        loadModel(outputId2, ItemModel)
+      ])
+        .then((items) => (
+          Promise.map(
+            items,
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching-correction.jsx]: Why put these args on separate lines?

File: web_client/analysis-modules/feature-matching-correction.jsx [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        ))
+        .then((elements) => { applyResultElements = elements; });
+    });
+};
+
+const main = (payload) => Promise.resolve(priorData = payload)
+  .then(() => D(actions.registerAnalysisAction(
+    'feature-match-correction', 'applyMatch', applyMatch
+  )))
+
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching-correction.jsx]: Blank line.

File: web_client/analysis-modules/feature-matching-correction.jsx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  .then(() => {
+    let element = priorData.page2ElementObjects[0];
+    element.baseMatching = priorData.matchResults.assignments;
+  })
+
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching-correction.jsx]: Blank line.

File: web_client/analysis-modules/feature-matching-correction.jsx [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+  .then(() => Promise.mapSeries(
+    priorData.page2ElementObjects,
+    (e) => D(actions.addAnalysisElement(e, priorData.page2))
+  ))
+  .then((elements) => {
+    priorData.page2Elements = elements;
+  })
+
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching-correction.jsx]: Blank line.

File: web_client/analysis-modules/feature-matching.jsx [hunk 12]
```diff
@@ -66,182 +42,51 @@ const computeMatch = (data, page) => {
...
-        applyTabElements,
-        (e) => D(actions.addAnalysisElement(e, applyTabElement))
-      ))
-      .then(([e]) => (matchChooserElement = e));
+  .then(() => analysisUtils.fetchAnalysisPage('feature-match-correction'))
+  .then(({ elements, ...page }) => {
+    page2ElementObjects = elements || [];
+    return D(actions.addAnalysisPage({ ...page, enabled: false }));
   })
+  .then((page) => (page2 = page))
```
Comment (waxlamp) [inline web_client/analysis-modules/feature-matching.jsx]: Remove parens around `page2 = page`.

**Review by waxlamp — APPROVED** [review]

File: server/task_specs/feature-match.py (hunk 2 above)
Comment (opadron) [reply server/task_specs/feature-match.py]: Yeah, that's right: this line should be left as-is.

**Pull request merged**